bindings/rust/src/lib.rs

```diff
@@ -4,8 +4,144 @@
+}
+
+pub struct Module {
+    ptr: NonNull<sys::FizzyModule>,
+}
+
+impl Drop for Module {
+    fn drop(&mut self) {
+        unsafe { sys::fizzy_free_module(self.ptr.as_ptr()) }
+    }
+}
+
+pub fn parse<T: AsRef<[u8]>>(input: &T) -> Result<Module, ()> {
+    let ptr = unsafe { sys::fizzy_parse(input.as_ref().as_ptr(), input.as_ref().len()) };
+    if ptr.is_null() {
+        return Err(());
+    }
+    Ok(Module {
+        ptr: unsafe { NonNull::new_unchecked(ptr) },
+    })
+}
+
+pub struct Instance {
+    ptr: NonNull<sys::FizzyInstance>,
+}
+
+impl Drop for Instance {
+    fn drop(&mut self) {
+        unsafe { sys::fizzy_free_instance(self.ptr.as_ptr()) }
+    }
+}
+
+impl Module {
+    // TODO: support imported functions{
+    pub fn instantiate(self) -> Result<Instance, ()> {
+        let ptr = unsafe { sys::fizzy_instantiate(self.ptr.as_ptr(), std::ptr::null_mut(), 0) };
+        // Forget Module (and avoid calling drop) because it has been consumed by instantiate (even if it failed).
+        core::mem::forget(self);
+        if ptr.is_null() {
+            return Err(());
+        }
+        Ok(Instance {
+            ptr: unsafe { NonNull::new_unchecked(ptr) },
+        })
+    }
+}
+
+// TODO: add a proper wrapper (enum type for Value, and Result<Value, Trap> for ExecutionResult)
+
+pub type Value = sys::FizzyValue;
+
+// NOTE: the union does not have i32
+impl Value {
+    pub fn as_i32(&self) -> i32 {
+        unsafe { self.i64 as i32 }
+    }
+    pub fn as_u32(&self) -> u32 {
+        unsafe { self.i64 as u32 }
+    }
+    pub fn as_i64(&self) -> i64 {
+        unsafe { self.i64 as i64 }
+    }
+    pub fn as_u64(&self) -> u64 {
+        unsafe { self.i64 }
+    }
+    pub fn as_f32(&self) -> f32 {
+        unsafe { self.f32 }
+    }
+    pub fn as_f64(&self) -> f64 {
+        unsafe { self.f64 }
+    }
+}
+
+impl From<i32> for Value {
+    fn from(v: i32) -> Self {
+        Value { i64: v as u64 }
+    }
+}
+
+impl From<u32> for Value {
+    fn from(v: u32) -> Self {
+        Value { i64: v as u64 }
+    }
+}
+
+impl From<i64> for Value {
+    fn from(v: i64) -> Self {
+        Value { i64: v as u64 }
+    }
+}
+
+impl From<u64> for Value {
+    fn from(v: u64) -> Self {
+        Value { i64: v }
+    }
+}
+impl From<f32> for Value {
+    fn from(v: f32) -> Self {
+        Value { f32: v }
+    }
+}
+
+impl From<f64> for Value {
+    fn from(v: f64) -> Self {
+        Value { f64: v }
+    }
+}
+
+pub struct ExecutionResult(sys::FizzyExecutionResult);
+
+impl ExecutionResult {
+    pub fn trapped(&self) -> bool {
+        self.0.trapped
+    }
+    pub fn value(&self) -> Option<Value> {
+        if self.0.has_value {
+            return Some(self.0.value);
+        }
+        None
+    }
+}
+
+impl From<ExecutionResult> for sys::FizzyExecutionResult {
+    fn from(v: ExecutionResult) -> Self {
+        v.0
+    }
+}
+
+impl Instance {
+    pub fn execute(&mut self, func_idx: u32, args: &[Value]) -> ExecutionResult {
```

Comment:
Should rename this `unsafe_execute` and make it unsafe. It can be used to squeeze out speed, at the cost of safety.